File: templates/python/bdapps_client.py
"""
bdapps API client — Python port of templates/typescript/bdapps-client.ts.

One `_post()` helper injects credentials, applies a timeout, and turns non-S1000
responses into typed errors. Every service is a thin wrapper that resolves its
endpoint through `config.require_endpoint()` — so calling an API your
application was not provisioned for fails locally with a clear message, rather
than as E1309 from the platform.

Standard library only, so it drops into any project without adding a dependency.
Using httpx or requests instead is fine and usually better in an existing
codebase — replace `_request()` and keep everything else:

    resp = httpx.post(url, json=payload, timeout=TIMEOUT_SECONDS)
    data = resp.json()          # note: still do NOT call resp.raise_for_status()
                                # as a success check — see _post().

SERVER-SIDE ONLY.
"""

from __future__ import annotations

import json
import re
import ssl
import urllib.error
import urllib.request
import uuid
from decimal import Decimal
from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence, Union

from bdapps_config import config

#: A single outbound call should never hang. Protocol constant, not config.
TIMEOUT_SECONDS = 15

#: bdapps hosts have served an incomplete certificate chain, which strict
#: clients reject. Do NOT "fix" that with ssl._create_unverified_context() or
#: verify=False — that lets anyone on the path read the applicationId and
#: password that can charge your subscribers. Supply the intermediate CA:
#:
#:     SSL_CONTEXT = ssl.create_default_context(cafile="certs/bdapps-chain.pem")
#:
#: See references/09-security-best-practices.md.
SSL_CONTEXT = ssl.create_default_context()

# ── Errors ───────────────────────────────────────────────────────────────────

#: Platform-side. Worth retrying with backoff.
TRANSIENT = frozenset(
    {
        "E1316", "E1318", "E1319", "E1332", "E1341",
        "E1360", "E1363", "E1364", "E1600", "E1601", "E1602", "E1603",
    }
)

#: Provisioning or credentials are wrong. Retrying will never help.
CONFIGURATION = frozenset(
    {
        "E1301", "E1302", "E1303", "E1304", "E1305", "E1306", "E1307",
        "E1309", "E1310", "E1311", "E1313", "E1315", "E1322", "E1323",
        "E1324", "E1327", "E1328", "E1329", "E1336", "E1371", "E1381",
        "E1383", "E1387",
    }
)

#: Codes meaning "the state you wanted already holds".
BENIGN_REGISTER = "E1351"    # user already registered
BENIGN_UNREGISTER = "E1356"  # user not registered
BENIGN_DEBIT = "E1379"       # transaction already completed

#: Mandatory on every CaaS call, with exactly one accepted value. It only ever
#: varies by being forgotten, so it is a constant rather than a caller argument.
PAYMENT_INSTRUMENT = "MobileAccount"


class BdappsError(Exception):
    def __init__(
        self,
        status_code: str,
        status_detail: str,
        service: str,
        raw: Optional[Mapping[str, Any]] = None,
    ) -> None:
        super().__init__(f"[{status_code}] {status_detail} ({service})")
        self.status_code = status_code
        self.status_detail = status_detail
        self.service = service
        self.raw = raw

    @property
    def retryable(self) -> bool:
        return self.status_code in TRANSIENT

    @property
    def is_configuration(self) -> bool:
        return self.status_code in CONFIGURATION


# ── Helpers ──────────────────────────────────────────────────────────────────

_SEPARATORS = re.compile(r"[\s()\-]")


def to_tel_address(msisdn: str) -> str:
    """
    Normalise a subscriber address. The ONLY place `tel:` is added.

    Accepts an already-prefixed address, a masked hash, +94…, 0094… or a local
    07… number.
    """
    trimmed = (msisdn or "").strip()
    if not trimmed:
        raise ValueError("[bdapps] Empty subscriber address")
    if trimmed.lower().startswith("tel:"):
        return trimmed

    digits = _SEPARATORS.sub("", trimmed).lstrip("+")
    if digits.startswith("00"):
        digits = digits[2:]
    if digits.startswith("0") and len(digits) == 10:
        digits = "94" + digits[1:]
    return f"tel:{digits}"


def mask_address(address: str) -> str:
    """Mask a subscriber address for logging. Never log the raw value."""
    body = re.sub(r"^tel:", "", address, flags=re.IGNORECASE)
    if len(body) <= 6:
        return "tel:***"
    return f"tel:{body[:3]}{'*' * (len(body) - 6)}{body[-3:]}"


def generate_external_trx_id() -> str:
    """A unique, persistable idempotency key for a charge. Max 32 characters."""
    return uuid.uuid4().hex


def format_amount(amount: Union[Decimal, str]) -> str:
    """
    Money crosses the wire as a string. Keep it in Decimal in your own code —
    a float will eventually charge someone 99.99999 rupees.
    """
    if isinstance(amount, float):  # pragma: no cover - guard against a real bug
        raise TypeError("[bdapps] Use Decimal or str for money, never float")
    return str(amount)


# ── Core ─────────────────────────────────────────────────────────────────────


def _request(url: str, body: bytes) -> Dict[str, Any]:
    request = urllib.request.Request(
        url,
        data=body,
        method="POST",
        headers={"Content-Type": "application/json", "Content-Length": str(len(body))},
    )
    try:
        with urllib.request.urlopen(
            request, timeout=TIMEOUT_SECONDS, context=SSL_CONTEXT
        ) as response:
            text = response.read().decode("utf-8")
    except urllib.error.HTTPError as exc:
        # bdapps answers 200 even for its own errors, so a non-2xx here is a
        # gateway/proxy problem — but the body may still be the real answer.
        text = exc.read().decode("utf-8", errors="replace")
    try:
        return json.loads(text)
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"[bdapps] Non-JSON response: {text[:200]}") from exc


def _post(
    service: str,
    url: str,
    body: Mapping[str, Any],
    benign_codes: Iterable[str] = (),
) -> Dict[str, Any]:
    payload = {
        "applicationId": config.application_id,
        "password": config.password,
        **body,
    }
    data = _request(url, json.dumps(payload).encode("utf-8"))

    status = data.get("statusCode")
    if status == "S1000" or status in set(benign_codes):
        return data
    raise BdappsError(
        str(status), str(data.get("statusDetail", "")), service, data
    )


# ── SMS ──────────────────────────────────────────────────────────────────────


def send_sms(
    to: Union[str, Sequence[str]],
    message: str,
    *,
    source_address: Optional[str] = None,
    delivery_status_request: Optional[str] = None,
    encoding: Optional[str] = None,
) -> Dict[str, Any]:
    """Send an MT SMS to one or more subscribers."""
    recipients: List[str] = [
        to_tel_address(x) for x in ([to] if isinstance(to, str) else list(to))
    ]
    if "tel:all" in recipients:
        raise ValueError(
            "[bdapps] Use broadcast_sms() for tel:all — broadcasts must be deliberate."
        )
    body: Dict[str, Any] = {"message": message, "destinationAddresses": recipients}
    if source_address:
        body["sourceAddress"] = source_address
    if delivery_status_request:
        body["deliveryStatusRequest"] = delivery_status_request
    if encoding:
        body["encoding"] = encoding
    return _post("sms-send", config.require_endpoint("sms_send"), body)


BROADCAST_CONFIRMATION = "I_HAVE_VERIFIED_THIS_GOES_TO_ALL_SUBSCRIBERS"


def broadcast_sms(message: str, confirmation: str, **options: Any) -> Dict[str, Any]:
    """
    Send to the ENTIRE subscribed base.

    Deliberately separate from send_sms so it can never be reached by accident —
    check the subscriber base size first, and put an authorisation check in
    front of this.
    """
    if confirmation != BROADCAST_CONFIRMATION:
        raise ValueError("[bdapps] Broadcast confirmation token missing")
    body: Dict[str, Any] = {"message": message, "destinationAddresses": ["tel:all"]}
    body.update({k: v for k, v in options.items() if v is not None})
    return _post("sms-send", config.require_endpoint("sms_send"), body)


# ── USSD ─────────────────────────────────────────────────────────────────────


def send_ussd(
    *, session_id: str, destination_address: str, message: str, operation: str
) -> Dict[str, Any]:
    """
    Send a USSD screen.

    `session_id` MUST be the one the platform sent you. Use "mt-fin" for the
    final screen — anything else leaves the session hanging until the network
    times out.
    """
    if operation not in {"mt-init", "mt-cont", "mt-fin"}:
        raise ValueError(f"[bdapps] Invalid ussdOperation '{operation}'")
    return _post(
        "ussd-send",
        config.require_endpoint("ussd_send"),
        {
            "message": message,
            "sessionId": session_id,
            "ussdOperation": operation,
            "destinationAddress": to_tel_address(destination_address),
            "encoding": "440",
            "version": "1.0",
        },
    )


# ── Subscription ─────────────────────────────────────────────────────────────


def register(subscriber_id: str) -> Dict[str, Any]:
    """
    Opt a subscriber in. Only call this with recorded, explicit consent.

    E1351 (already registered) is accepted as success — the desired state holds.
    """
    return _post(
        "subscription-register",
        config.require_endpoint("subscription_send"),
        {
            "subscriberId": to_tel_address(subscriber_id),
            "action": "1",
            "version": "1.0",
        },
        [BENIGN_REGISTER],
    )


def unregister(subscriber_id: str) -> Dict[str, Any]:
    """
    Opt a subscriber out.

    E1356 (not registered) is accepted as success — the desired state holds.
    """
    return _post(
        "subscription-unregister",
        config.require_endpoint("subscription_send"),
        {
            "subscriberId": to_tel_address(subscriber_id),
            "action": "0",
            "version": "1.0",
        },
        [BENIGN_UNREGISTER],
    )


def get_subscription_status(subscriber_id: str) -> Dict[str, Any]:
    """Check one subscriber's status. For reconciliation, not per-request gating."""
    return _post(
        "subscription-status",
        config.require_endpoint("subscription_status"),
        {"subscriberId": to_tel_address(subscriber_id)},
    )


def query_base() -> Dict[str, Any]:
    """
    Subscriber base size. Needs no subscriber and charges nothing, which also
    makes it the best connectivity and credential smoke test.

    `baseSize` comes back as a string, so a parsed integer is added alongside.
    """
    response = _post(
        "subscription-query-base",
        config.require_endpoint("subscription_query_base"),
        {},
    )
    return {**response, "size": int(response.get("baseSize") or 0)}


# ── OTP ──────────────────────────────────────────────────────────────────────


def request_otp(
    *,
    subscriber_id: str,
    meta_data: Mapping[str, Any],
    application_hash: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Send an OTP to a plain mobile number.

    Rate-limit per number AND per IP before calling, or the app becomes an
    SMS-bombing tool. Keep the returned referenceNo server-side; never log it.
    """
    body: Dict[str, Any] = {
        "subscriberId": to_tel_address(subscriber_id),
        "applicationMetaData": dict(meta_data),
    }
    if application_hash:
        body["applicationHash"] = application_hash
    return _post("otp-request", config.require_endpoint("otp_request"), body)


def verify_otp(*, reference_no: str, otp: str) -> Dict[str, Any]:
    """
    Verify an OTP. Valid 60 minutes, maximum 3 attempts — enforce those limits
    on your side too. The returned subscriberId is the masked identifier to use
    for every subsequent call.
    """
    return _post(
        "otp-verify",
        config.require_endpoint("otp_verify"),
        {"referenceNo": reference_no, "otp": otp},
    )


# ── CaaS ─────────────────────────────────────────────────────────────────────


def debit(
    *,
    subscriber_id: str,
    amount: Union[Decimal, str],
    external_trx_id: str,
    currency: str = "BDT",
    account_id: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Charge a subscriber's mobile account.

    THIS MOVES REAL MONEY.

    - `external_trx_id` is your idempotency key. Generate it with
      generate_external_trx_id(), PERSIST IT, then call this.
    - There are deliberately no retries here. A timeout does NOT mean the charge
      failed. Resolve unknown outcomes by re-calling with the SAME
      external_trx_id — the platform de-duplicates on it.
      Never re-roll the id.
    - E1379 (already completed) is accepted as success.
    """
    if not external_trx_id:
        raise ValueError(
            "[bdapps] external_trx_id is required and must be persisted first"
        )
    if len(external_trx_id) > 32:
        raise ValueError("[bdapps] external_trx_id must be 32 characters or fewer")

    body: Dict[str, Any] = {
        "externalTrxId": external_trx_id,
        "subscriberId": to_tel_address(subscriber_id),
        "paymentInstrumentName": PAYMENT_INSTRUMENT,
        "amount": format_amount(amount),
        "currency": currency,
    }
    if account_id:
        body["accountId"] = account_id

    return _post(
        "caas-direct-debit",
        config.require_endpoint("caas_debit"),
        body,
        [BENIGN_DEBIT],
    )


def query_balance(
    *,
    subscriber_id: str,
    currency: str = "BDT",
    account_id: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Query chargeable balance.

    Advisory only: the balance can change between this and the debit. Always
    handle E1378 on the debit regardless of what this returned.

    Not configuring BDAPPS_CAAS_BALANCE_URL is how you disable this on an
    application without the "Enable Query Balance Requests" toggle.
    """
    body: Dict[str, Any] = {
        "subscriberId": to_tel_address(subscriber_id),
        "paymentInstrumentName": PAYMENT_INSTRUMENT,
        "currency": currency,
    }
    if account_id:
        body["accountId"] = account_id
    return _post("caas-balance-query", config.require_endpoint("caas_balance"), body)


# ── Extension point ──────────────────────────────────────────────────────────
#
# Adding a service bdapps publishes later:
#
#   1. Add its URL variable to .env.example and to _ENDPOINT_VARS in
#      bdapps_config.py
#   2. Add one wrapper here:
#
#          def new_thing(**kwargs):
#              return _post("new-thing", config.require_endpoint("new_thing"), kwargs)
#
# It inherits credential injection, the timeout, error mapping and the
# not-provisioned guard for free. Do not build a parallel client.

File: templates/php/BdappsClient.php
<?php

declare(strict_types=1);

namespace App\bdapps;

use InvalidArgumentException;
use RuntimeException;

/**
 * bdapps API client — PHP port of templates/typescript/bdapps-client.ts.
 *
 * One post() helper injects credentials, applies a timeout, and turns non-S1000
 * responses into typed errors. Every service is a thin wrapper that resolves its
 * endpoint through BdappsConfig::requireEndpoint() — so calling an API your
 * application was not provisioned for fails locally with a clear message, rather
 * than as E1309 from the platform.
 *
 * Uses the cURL extension so it drops into any project without Composer
 * dependencies. Guzzle is a fine substitute — replace request() and keep
 * everything else, but do NOT enable http_errors as a success check: bdapps
 * returns HTTP 200 for its own failures.
 *
 * Requires PHP 8.1+ (readonly properties, enums-free but typed constants). BdappsException
 * shares this file for readability — split it into its own file if you autoload with PSR-4.
 *
 * SERVER-SIDE ONLY.
 */
final class BdappsClient
{
    /** A single outbound call should never hang. Protocol constant, not config. */
    private const TIMEOUT_SECONDS = 15;

    /** Platform-side. Worth retrying with backoff. */
    public const TRANSIENT = [
        'E1316', 'E1318', 'E1319', 'E1332', 'E1341',
        'E1360', 'E1363', 'E1364', 'E1600', 'E1601', 'E1602', 'E1603',
    ];

    /** Provisioning or credentials are wrong. Retrying will never help. */
    public const CONFIGURATION = [
        'E1301', 'E1302', 'E1303', 'E1304', 'E1305', 'E1306', 'E1307',
        'E1309', 'E1310', 'E1311', 'E1313', 'E1315', 'E1322', 'E1323',
        'E1324', 'E1327', 'E1328', 'E1329', 'E1336', 'E1371', 'E1381',
        'E1383', 'E1387',
    ];

    /** Codes meaning "the state you wanted already holds". */
    public const BENIGN_REGISTER   = 'E1351'; // user already registered
    public const BENIGN_UNREGISTER = 'E1356'; // user not registered
    public const BENIGN_DEBIT      = 'E1379'; // transaction already completed

    /**
     * Mandatory on every CaaS call, with exactly one accepted value. It only ever
     * varies by being forgotten, so it is a constant rather than a caller argument.
     */
    public const PAYMENT_INSTRUMENT = 'MobileAccount';

    public const BROADCAST_CONFIRMATION = 'I_HAVE_VERIFIED_THIS_GOES_TO_ALL_SUBSCRIBERS';

    public function __construct(private readonly BdappsConfig $config)
    {
    }

    /* ── Helpers ──────────────────────────────────────────────────────────── */

    /**
     * Normalise a subscriber address. The ONLY place `tel:` is added.
     *
     * Accepts an already-prefixed address, a masked hash, +94…, 0094… or a
     * local 07… number.
     */
    public static function toTelAddress(string $msisdn): string
    {
        $trimmed = trim($msisdn);
        if ($trimmed === '') {
            throw new InvalidArgumentException('[bdapps] Empty subscriber address');
        }
        if (stripos($trimmed, 'tel:') === 0) {
            return $trimmed;
        }

        $digits = ltrim((string) preg_replace('/[\s()\-]/', '', $trimmed), '+');
        if (str_starts_with($digits, '00')) {
            $digits = substr($digits, 2);
        }
        if (str_starts_with($digits, '0') && strlen($digits) === 10) {
            $digits = '94' . substr($digits, 1);
        }

        return 'tel:' . $digits;
    }

    /** Mask a subscriber address for logging. Never log the raw value. */
    public static function maskAddress(string $address): string
    {
        $body = (string) preg_replace('/^tel:/i', '', $address);
        if (strlen($body) <= 6) {
            return 'tel:***';
        }

        return 'tel:' . substr($body, 0, 3)
            . str_repeat('*', strlen($body) - 6)
            . substr($body, -3);
    }

    /** A unique, persistable idempotency key for a charge. Max 32 characters. */
    public static function generateExternalTrxId(): string
    {
        return bin2hex(random_bytes(16));
    }

    /* ── Core ─────────────────────────────────────────────────────────────── */

    /**
     * @param array<string, mixed> $body
     * @param list<string>         $benignCodes
     *
     * @return array<string, mixed>
     */
    private function post(string $service, string $url, array $body, array $benignCodes = []): array
    {
        $payload = json_encode(
            array_merge(
                [
                    'applicationId' => $this->config->applicationId,
                    'password'      => $this->config->password,
                ],
                $body
            ),
            JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES
        );

        $data = $this->request($service, $url, $payload);

        $statusCode = (string) ($data['statusCode'] ?? '');
        if ($statusCode === 'S1000' || in_array($statusCode, $benignCodes, true)) {
            return $data;
        }

        throw new BdappsException(
            $statusCode,
            (string) ($data['statusDetail'] ?? ''),
            $service,
            $data
        );
    }

    /** @return array<string, mixed> */
    private function request(string $service, string $url, string $payload): array
    {
        $handle = curl_init($url);
        if ($handle === false) {
            throw new RuntimeException("[bdapps] {$service}: could not initialise cURL");
        }

        curl_setopt_array($handle, [
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => $payload,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => self::TIMEOUT_SECONDS,
            CURLOPT_HTTPHEADER     => ['Content-Type: application/json'],
            // bdapps hosts have served an incomplete certificate chain. Do NOT
            // "fix" that with CURLOPT_SSL_VERIFYPEER => false — that lets anyone
            // on the path read the applicationId and password that can charge
            // your subscribers. Supply the intermediate CA instead:
            //   CURLOPT_CAINFO => __DIR__ . '/certs/bdapps-chain.pem',
            CURLOPT_SSL_VERIFYPEER => true,
            CURLOPT_SSL_VERIFYHOST => 2,
        ]);

        $raw = curl_exec($handle);
        $error = curl_error($handle);
        curl_close($handle);

        if ($raw === false) {
            throw new RuntimeException("[bdapps] {$service}: transport failure: {$error}");
        }

        // The HTTP status is deliberately not consulted: bdapps returns 200 for
        // application-level failures, and the real outcome is statusCode.
        $decoded = json_decode((string) $raw, true);
        if (!is_array($decoded)) {
            throw new RuntimeException(
                "[bdapps] {$service}: non-JSON response: " . substr((string) $raw, 0, 200)
            );
        }

        return $decoded;
    }

    /* ── SMS ──────────────────────────────────────────────────────────────── */

    /**
     * Send an MT SMS to one or more subscribers.
     *
     * @param string|list<string>  $to
     * @param array<string, mixed> $options sourceAddress, deliveryStatusRequest,
     *                                      encoding, binaryHeader
     *
     * @return array<string, mixed>
     */
    public function sendSms(string|array $to, string $message, array $options = []): array
    {
        $recipients = array_map(
            [self::class, 'toTelAddress'],
            is_array($to) ? $to : [$to]
        );
        if (in_array('tel:all', $recipients, true)) {
            throw new InvalidArgumentException(
                '[bdapps] Use broadcastSms() for tel:all — broadcasts must be deliberate.'
            );
        }

        return $this->post('sms-send', $this->config->requireEndpoint('smsSend'), array_merge(
            ['message' => $message, 'destinationAddresses' => array_values($recipients)],
            $options
        ));
    }

    /**
     * Send to the ENTIRE subscribed base.
     *
     * Deliberately separate from sendSms() so it can never be reached by
     * accident — check the subscriber base size first, and put an authorisation
     * check in front of this.
     *
     * @param array<string, mixed> $options
     *
     * @return array<string, mixed>
     */
    public function broadcastSms(string $message, string $confirmation, array $options = []): array
    {
        if ($confirmation !== self::BROADCAST_CONFIRMATION) {
            throw new InvalidArgumentException('[bdapps] Broadcast confirmation token missing');
        }

        return $this->post('sms-send', $this->config->requireEndpoint('smsSend'), array_merge(
            ['message' => $message, 'destinationAddresses' => ['tel:all']],
            $options
        ));
    }

    /* ── USSD ─────────────────────────────────────────────────────────────── */

    /**
     * Send a USSD screen.
     *
     * $sessionId MUST be the one the platform sent you. Use 'mt-fin' for the
     * final screen — anything else leaves the session hanging until the network
     * times out.
     *
     * @return array<string, mixed>
     */
    public function sendUssd(
        string $sessionId,
        string $destinationAddress,
        string $message,
        string $operation
    ): array {
        if (!in_array($operation, ['mt-init', 'mt-cont', 'mt-fin'], true)) {
            throw new InvalidArgumentException("[bdapps] Invalid ussdOperation '{$operation}'");
        }

        return $this->post('ussd-send', $this->config->requireEndpoint('ussdSend'), [
            'message'            => $message,
            'sessionId'          => $sessionId,
            'ussdOperation'      => $operation,
            'destinationAddress' => self::toTelAddress($destinationAddress),
            'encoding'           => '440',
            'version'            => '1.0',
        ]);
    }

    /* ── Subscription ─────────────────────────────────────────────────────── */

    /**
     * Opt a subscriber in. Only call this with recorded, explicit consent.
     *
     * E1351 (already registered) is accepted as success — the desired state holds.
     *
     * @return array<string, mixed>
     */
    public function register(string $subscriberId): array
    {
        return $this->post(
            'subscription-register',
            $this->config->requireEndpoint('subscriptionSend'),
            [
                'subscriberId' => self::toTelAddress($subscriberId),
                'action'       => '1',
                'version'      => '1.0',
            ],
            [self::BENIGN_REGISTER]
        );
    }

    /**
     * Opt a subscriber out.
     *
     * E1356 (not registered) is accepted as success — the desired state holds.
     *
     * @return array<string, mixed>
     */
    public function unregister(string $subscriberId): array
    {
        return $this->post(
            'subscription-unregister',
            $this->config->requireEndpoint('subscriptionSend'),
            [
                'subscriberId' => self::toTelAddress($subscriberId),
                'action'       => '0',
                'version'      => '1.0',
            ],
            [self::BENIGN_UNREGISTER]
        );
    }

    /**
     * Check one subscriber's status. For reconciliation, not per-request gating.
     *
     * @return array<string, mixed>
     */
    public function subscriptionStatus(string $subscriberId): array
    {
        return $this->post(
            'subscription-status',
            $this->config->requireEndpoint('subscriptionStatus'),
            ['subscriberId' => self::toTelAddress($subscriberId)]
        );
    }

    /**
     * Subscriber base size. Needs no subscriber and charges nothing, which also
     * makes it the best connectivity and credential smoke test.
     */
    public function queryBase(): int
    {
        $data = $this->post(
            'subscription-query-base',
            $this->config->requireEndpoint('subscriptionQueryBase'),
            []
        );

        return (int) ($data['baseSize'] ?? 0); // documented as a string
    }

    /* ── OTP ──────────────────────────────────────────────────────────────── */

    /**
     * Send an OTP to a plain mobile number.
     *
     * Rate-limit per number AND per IP before calling, or the app becomes an
     * SMS-bombing tool. Keep the returned referenceNo server-side; never log it.
     *
     * @param array<string, mixed> $applicationMetaData
     *
     * @return array<string, mixed>
     */
    public function requestOtp(string $subscriberId, array $applicationMetaData): array
    {
        return $this->post('otp-request', $this->config->requireEndpoint('otpRequest'), [
            'subscriberId'        => self::toTelAddress($subscriberId),
            'applicationMetaData' => $applicationMetaData,
        ]);
    }

    /**
     * Verify an OTP. Valid 60 minutes, maximum 3 attempts — enforce those limits
     * on your side too. The returned subscriberId is the masked identifier to
     * use for every subsequent call.
     *
     * @return array<string, mixed>
     */
    public function verifyOtp(string $referenceNo, string $otp): array
    {
        return $this->post('otp-verify', $this->config->requireEndpoint('otpVerify'), [
            'referenceNo' => $referenceNo,
            'otp'         => $otp,
        ]);
    }

    /* ── CaaS ─────────────────────────────────────────────────────────────── */

    /**
     * Charge a subscriber's mobile account.
     *
     * THIS MOVES REAL MONEY.
     *
     * - $externalTrxId is your idempotency key. Generate it with
     *   generateExternalTrxId(), PERSIST IT, then call this.
     * - There are deliberately no retries here. A timeout does NOT mean the
     *   charge failed. Resolve unknown outcomes by re-calling with the SAME
     *   externalTrxId — the platform de-duplicates on it.
     *   Never re-roll the id.
     * - E1379 (already completed) is accepted as success.
     * - $amount is a string: keep money in bcmath or integer minor units. A
     *   float will eventually charge someone 99.99999 rupees.
     *
     * @return array<string, mixed>
     */
    public function debit(
        string $subscriberId,
        string $amount,
        string $externalTrxId,
        string $currency = 'BDT'
    ): array {
        if ($externalTrxId === '') {
            throw new InvalidArgumentException(
                '[bdapps] externalTrxId is required and must be persisted first'
            );
        }
        if (strlen($externalTrxId) > 32) {
            throw new InvalidArgumentException(
                '[bdapps] externalTrxId must be 32 characters or fewer'
            );
        }

        return $this->post(
            'caas-direct-debit',
            $this->config->requireEndpoint('caasDebit'),
            [
                'externalTrxId'         => $externalTrxId,
                'subscriberId'          => self::toTelAddress($subscriberId),
                'paymentInstrumentName' => self::PAYMENT_INSTRUMENT,
                'amount'                => $amount,
                'currency'              => $currency,
            ],
            [self::BENIGN_DEBIT]
        );
    }

    /**
     * Query chargeable balance.
     *
     * Advisory only: the balance can change between this and the debit. Always
     * handle E1378 on the debit regardless of what this returned.
     *
     * @return array<string, mixed>
     */
    public function queryBalance(string $subscriberId, string $currency = 'BDT'): array
    {
        return $this->post('caas-balance-query', $this->config->requireEndpoint('caasBalance'), [
            'subscriberId'          => self::toTelAddress($subscriberId),
            'paymentInstrumentName' => self::PAYMENT_INSTRUMENT,
            'currency'              => $currency,
        ]);
    }

    /* ── Extension point ───────────────────────────────────────────────────
     *
     * Adding a service bdapps publishes later:
     *
     *   1. Add its URL variable to .env.example and to ENDPOINT_VARS in
     *      BdappsConfig
     *   2. Add one wrapper here that calls $this->post() with the new key.
     *
     * It inherits credential injection, the timeout, error mapping and the
     * not-provisioned guard for free. Do not build a parallel client.
     */
}

/** A non-S1000 application-level response. */
final class BdappsException extends RuntimeException
{
    /** @param array<string, mixed> $raw */
    public function __construct(
        public readonly string $statusCode,
        public readonly string $statusDetail,
        public readonly string $service,
        public readonly array $raw = [],
    ) {
        parent::__construct("[{$statusCode}] {$statusDetail} ({$service})");
    }

    public function isRetryable(): bool
    {
        return in_array($this->statusCode, BdappsClient::TRANSIENT, true);
    }

    public function isConfiguration(): bool
    {
        return in_array($this->statusCode, BdappsClient::CONFIGURATION, true);
    }
}
